Objects/moduleobject.c

```diff
@@ -679,12 +679,27 @@ module_repr(PyModuleObject *m)
 static PyObject*
 module_getattro(PyModuleObject *m, PyObject *name)
 {
-    PyObject *attr, *mod_name;
+    PyObject *attr, *mod_name, *getattr;
+    PyObject* stack[1] = {name};
     attr = PyObject_GenericGetAttr((PyObject *)m, name);
-    if (attr || !PyErr_ExceptionMatches(PyExc_AttributeError))
+    if (attr || !PyErr_ExceptionMatches(PyExc_AttributeError)) {
         return attr;
+    }
     PyErr_Clear();
     if (m->md_dict) {
+        _Py_IDENTIFIER(__getattr__);
+        getattr = _PyDict_GetItemId(m->md_dict, &PyId___getattr__);
+        if (getattr) {
+            if (!PyCallable_Check(getattr)) {
+                PyErr_SetString(PyExc_TypeError,
+                                "module __getattr__ must be callable");
+                return NULL;
+            }
+            return _PyObject_FastCall(getattr, stack, 1);
```

Comment:
Could you please test the following case?
```python
def __getattr__(name):
    if name != 'delgetattr':
        raise AttributeError
    del globals()['__getattr__']
    # do something more
```
I suspect that it can crash (but not sure).

Reply:
Thanks for review! I think I addressed all the comments, but there is a little problem. This test that I added as ``test_module_getattr_tricky`` fails with ``AttributeError`` in refleak hunting mode ``-R 3:3`` (it passes in normal mode). Presumably this is because the module stays in memory between sequential runs and ``__getattr__`` is already deleted after the first test. How do I fix this?

Reply:
If the test didn't crash then perhaps we shouldn't add it.

Reply:
I think it makes sense to keep something like this just in case. I simplified the test (``__getattr__`` always raises ``AttributeError``, but will also delete itself). This way it works also in refleak hunting mode.